src/MetaPaymaster.sol

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.20;
+
+/* solhint-disable reason-string */
+
+import "@account-abstraction/interfaces/IEntryPoint.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+contract MetaPaymaster is OwnableUpgradeable {
+    IEntryPoint public entryPoint;
+    mapping(address => uint256) public balances;
+    uint256 public total;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(address owner, IEntryPoint _entryPoint) public initializer {
+        __Ownable_init(owner);
+        entryPoint = _entryPoint;
+    }
+
+    function fund(uint256 actualGasCost) external returns (bool) {
+        if (address(this).balance < actualGasCost) {
+            return false;
+        }
+        if (balances[msg.sender] < actualGasCost) {
+            return false;
+        }
+        total -= actualGasCost;
+        balances[msg.sender] -= actualGasCost;
+        entryPoint.depositTo{value: actualGasCost}(msg.sender);
+        return true;
+    }
+
+    function balance() public view returns (uint256) {
+        return balances[msg.sender];
+    }
+
+    function balanceOf(address account) public view returns (uint256) {
+        return balances[account];
+    }
+
+    function depositTo(address account) public payable {
+        total += msg.value;
+        balances[account] += msg.value;
+    }
+
+    function setBalance(address account, uint256 amount) public onlyOwner {
```

Comment:
how do we imagine using this? Allows us to give credits without actually paying?

Reply:
Decoupling the balance per paymaster from the contract balance allows undercollateralization. In the future we may see lots of paymasters using this metapaymaster, it seems ideal to have the ability to spend from a pool of funds, rather than backing the balances 100%.

Reply:
yeah I think makes sense